### PR TITLE
Vision API added

### PR DESCRIPTION
**Pull request by jakethekoenig**
Type `/screenshot path` where path can be a url or local file and mentat will open it with selenium and take a screenshot and put it in the LLMs context. You can see it with `/conversation`.

* A Message class is introduced which holds onto the role, text and optional image path. It handles representing it in the form the llm wants.
* The conversation viewer no longer puts the messages in pre so you don't have to side scroll them anymore. New lines in messages are replaced with \<br\>.
* The new transcript logging format is breaking with the old one so you'll have to clear out your logs before you use this.

Todo:
- [x] Automatically switch the user to `gpt-vision-preview` with a warning when they run screenshot.
- [x] `/screenshot` should work with no arguments and screenshot the open browser.
- [x] Tests!

Long term
* Because we're using selenium we could allow the LLM to interact with the DOM in more complex ways. We could let it request parts of the DOM. We could let it navigate the page and explore.

### CODE REVIEW

File: tests/license_check.py
Comment (PCSwingle) [inline tests/license_check.py:21]: I think you'll need to add just BSD (without the 'License' at the end)

File: mentat/commands.py
Comment (PCSwingle) [inline mentat/commands.py:396]: Let's not raise a UserError; these should be used to exit mentat with an error message; in this case, we just want to fail the command

File: mentat/message.py
Comment (PCSwingle) [inline mentat/message.py:17]: Since none of this will be necessary once I get the OpenAI API 1.0 PR through, I'm wondering if we should hold off on this PR until then?

File: mentat/vision/vision_manager.py
Comment (PCSwingle) [inline mentat/vision/vision_manager.py:18]: How about we change the name of this function to _open_browser?

File: mentat/vision/vision_manager.py
Comment (PCSwingle) [inline mentat/vision/vision_manager.py:24]: Does this open a new browser every time and does it close the browser afterwards or not?

**Comment by PCSwingle** [comment]
Looks great to me! I left a few comments. I'm mostly wondering if we want to delay merging this until OpenAI gets updated; I'm not entirely sure how I feel and I'd be ok with either. The only downside to merging this early is that we'd have to go ahead and remove all of the Message stuff, but that would be doable.

File: mentat/llm_api.py
Comment (jakethekoenig) [reply mentat/llm_api.py:120]: Oh yeah it's sort of weird: The open ai api returns a max of 30 tokens by default if you include an image. I spent a long time debugging thinking it must be my problem. Not sure why they added that behavior.

File: mentat/message.py
Comment (jakethekoenig) [reply mentat/message.py:17]: Yeah I agree :( I should have looked at the API before doing this.

File: mentat/vision/vision_manager.py
Comment (jakethekoenig) [reply mentat/vision/vision_manager.py:24]: It opens in the same browser. The browser is automatically closed when mentat quits (or the user can close it themselves).

**Comment by jakethekoenig** [comment]
> Looks great to me! I left a few comments. I'm mostly wondering if we want to delay merging this until OpenAI gets updated; I'm not entirely sure how I feel and I'd be ok with either. The only downside to merging this early is that we'd have to go ahead and remove all of the Message stuff, but that would be doable.

I'm down to delay until after the OpenAI update. I still have to write tests for this and stuff so it isn't that close to being done. And I don't know how similar my Message class is to the one OpenAI has. I don't want to write code in the conversation_viewer that relies on my abstraction that you'd have to then rewrite.

File: mentat/llm_api.py
Comment (jakethekoenig) [inline mentat/llm_api.py:182]: Note to self: Need to round up not down here.

File: mentat/resources/templates/conversation_viewer.jinja
Comment (jakethekoenig) [inline mentat/resources/templates/conversation_viewer.jinja:33]: There's some code duplication here. I tried to use mentat to make a macro to fix it and it failed. I tried myself and it seemed like the arguments to the macro were passed as plaintext. I don't think it's that important to fix but if anyone who has experience with jinja macros it'd be nice.

File: mentat/llm_api_handler.py
Comment (jakethekoenig) [inline mentat/llm_api_handler.py:78]: There are a lot of ignores here. I think Pyright should have enough information given the classes to know entry is of type `ChatCompletionContentPartParam`. Maybe it would help if instead of iterating over items we just accessed the content which should be the only source of tokens anyway. For our use case the only attributes should be role and content. If we start using function calls there could be others.

I'm not sure as written it's exactly correct because when we iterate over items one of them will be the role which should already be counted in line 72. In my `llm_api_handler_test` line 33 I add 5 at the end but I think adding 4 would be correct.

**Review by jakethekoenig — COMMENTED** [review]
This PR is merged to use OpenAI's types now. Thanks again @PCSwingle. It makes the change a lot tighter and hopefully easier to review. Also the jinja refactor means it's no longer necessary to keep track of local image files for screenshots because we can inline the base64 encoded images.

I left two inline comments that would be nice to address but I don't think they're blocking.

I'm also not sure if we want to write an end to end screenshot test that actually fires up selenium and takes a screenshot. Github action runners don't have chrome by default and I'm worried it would slow down our test suite a lot.

File: mentat/llm_api.py
Comment (PCSwingle) [reply mentat/llm_api.py]: Huh, that is weird

File: mentat/commands.py
Comment (PCSwingle) [inline mentat/commands.py:408]: Maybe just say version of gpt? Could be using gpt-3.5

Comment (jakethekoenig) [reply mentat/commands.py:408]: Running the screenshot command doesn't call the model it just adds a screenshot to the conversation. The model will need to be `gpt-4-vision-preview` when the user next enters a prompt (and on all future prompts unless the run `/clear`). This is part of the reason I think this change of the model is safe: If the user doesn't want to use `gpt-4-vision-preview` they'll see the model is changed and they'll have the opportunity to change it back or quit before it is used.

File: mentat/llm_api_handler.py
Comment (PCSwingle) [reply mentat/llm_api_handler.py]: It could be a list of ToolCalls (in ChatCompletionAssistantMessageParam). I just added a commit that casts value as long as the key is "content".

 I think the +4 is for the im_start and im_end, not for the role, and the role is always one token which bumps it up to 5. I could be wrong though; I'll try curling some messages to openai (without streaming, so they'll give me the exact tokens) and running them through our function and see if I'm wrong.

File: mentat/llm_api_handler.py
Comment (jakethekoenig) [reply mentat/llm_api_handler.py:197]: Correct. I believe setting this parameter should have no effect mostly except for vision models where it overrides the max of 30. e.g. there's no concern the model will see this large value and try to fill up the space.

File: mentat/transcripts.py
Comment (jakethekoenig) [reply mentat/transcripts.py:12]: This actually does change how the prior messages will appear in the text transcript. e.g. I see prior messages like: `{"role": "user", "content": [{"type": "text", "text": "Describe the doodle"}]}` whereas previously the content would have been a string. This isn't breaking with previous transcipts because the jinja template handles both. I'm actually not 100% sure openai treats both the same. While you're curling could you double check that that message and `{"role": "user", "content": "Describe the doodle"}` return the same token count?

File: mentat/resources/templates/conversation_viewer.jinja
Comment (PCSwingle) [reply mentat/resources/templates/conversation_viewer.jinja:33]: I'll work on this

**Comment by PCSwingle** [comment]
TODO before merging:
Curl and check that our token counts are correct
Check that token counts are the same for list of content vs str content
/screenshot an html file isn't working for me; I think we need an absolute path and/or triple /// after file.

File: mentat/vision/vision_manager.py
Comment (jakethekoenig) [inline mentat/vision/vision_manager.py:32]: Good catch! I think we should have both. As is `/screenshot ~/.mentat/conversation_viewer.html` won't work.

Comment (PCSwingle) [reply mentat/vision/vision_manager.py:32]: Oh I didn't realize abspath didn't do that, thanks! I'll do that

File: mentat/resources/templates/conversation_viewer.jinja
Comment (PCSwingle) [reply mentat/resources/templates/conversation_viewer.jinja:1]: Hmm, maybe? That is weird

**Review by PCSwingle — APPROVED** [review]
Looks great! Thanks so much for adding this!

**Pull request merged**
